src/language-js/print/arrow-function.js

```diff
@@ -0,0 +1,335 @@
+import {
+  printDanglingComments,
+  printCommentsSeparately,
+} from "../../main/comments/print.js";
+import getNextNonSpaceNonCommentCharacterIndex from "../../utils/get-next-non-space-non-comment-character-index.js";
+import {
+  line,
+  softline,
+  group,
+  indent,
+  ifBreak,
+  join,
+  indentIfBreak,
+} from "../../document/builders.js";
+import { removeLines, willBreak } from "../../document/utils.js";
+import { ArgExpansionBailout } from "../../common/errors.js";
+import {
+  getFunctionParameters,
+  hasLeadingOwnLineComment,
+  isJsxElement,
+  isTemplateOnItsOwnLine,
+  shouldPrintComma,
+  startsWithNoLookaheadToken,
+  isBinaryish,
+  hasComment,
+  isCallLikeExpression,
+  isArrayOrTupleExpression,
+  isObjectOrRecordExpression,
+} from "../utils/index.js";
+import { locEnd } from "../loc.js";
+import { printFunctionParameters } from "./function-parameters.js";
+import { printReturnType, shouldPrintParamsWithoutParens } from "./function.js";
+
+/**
+ * @typedef {import("../../common/ast-path.js").default} AstPath
+ * @typedef {import("../../document/builders.js").Doc} Doc
+ */
+
+// In order to avoid confusion between
+// a => a ? a : a
+// a <= a ? a : a
+const shouldAddParensIfNotBreakCache = new WeakMap();
+function shouldAddParensIfNotBreak(node) {
+  if (!shouldAddParensIfNotBreakCache.has(node)) {
+    shouldAddParensIfNotBreakCache.set(
+      node,
+      node.type === "ConditionalExpression" &&
+        !startsWithNoLookaheadToken(
+          node,
+          (node) => node.type === "ObjectExpression"
+        )
+    );
+  }
+  return shouldAddParensIfNotBreakCache.get(node);
+}
+
+// We handle sequence expressions as the body of arrows specially,
+// so that the required parentheses end up on their own lines.
+const shouldAlwaysAddParens = (node) => node.type === "SequenceExpression";
+
+const isArrowFunctionChain = (node) =>
+  node.body.type === "ArrowFunctionExpression";
```

Comment:
[Question]
As of main, it looked to `currentNode.body.type !== "ArrowFunctionExpression" || args.expandLastArg` reading just `rec()` in `printArrowFunction()`.
Do the difference cause a problem?

Though it doesn't cause any problem as far as I tried, I'm not sure about it because I'm not familiar with `expandLastArg`.
Example of input I tried:
```js
fooooooooooooooooooooooo(action => a => b => 
  (dispatch123456789098765432123451234567789(action), 1, 2)
);
```

Reply:
> Do the difference cause a problem?

I also worried about it, and it may cause problem, but I forgot about it...

Related logic here  https://github.com/prettier/prettier/blob/7f292bd8f3cda8eade91012b50ab481b32e3d2e8/src/language-js/print/call-arguments.js#L221

Reply:
As my first impression, removing `|| args.expandLastArg` int `rec()` looked to make sense.
But it will change some results, and previous format looks slightly better.
I'm still investigating.
```diff
- request.get(
-   "https://preview-9992--prettier.netlify.app",
-   (head) =>
-     (body) =>
-     (modyLoremIpsumDolorAbstractProviderFactoryServiceModule) => {
-       console.log(head, body);
-     },
- );
+ request.get("https://preview-9992--prettier.netlify.app", (head) =>
+   (body) =>
+   (modyLoremIpsumDolorAbstractProviderFactoryServiceModule) => {
+     console.log(head, body);
+   });
```

Reply:
This yields same results regardless of `|| args.expandLastArg`. So `expandLastArg` looks to work.
```js
request.get("https://preview-9992--prettier.netlify.app", (head) => (body) => {
  console.log(head, body);
})
```

Reply:
I read the code again, when `expandLastArg` is true. `functionBody` can be an `ArrowFunctionExpression`, and `functionBody` decides how the body should be printed. Can't find a case to test. [`98c36c2` (#14633)](https://github.com/prettier/prettier/pull/14633/commits/98c36c2e881ffb328dcdec194f35a0f5e2af33b0) should restore the old behavior, I think.

Reply:
The following code can make `functionBody` be an `ArrowFunctionExpression` and effect `shouldBreakChain`

```
a("", "", () => ({}) => {
  a;
});
```

- last argument is `ArrrowFunctionExpression` satisfies `couldExpandArg`
- first two arguments satisfies `args.length !== 2` in `shouldExpandLastArg`
- first two arguments is very simple so they will pass `headArgs.some(willBreak)` in `printCallArguments`

Reply:
Finally find a case formats differently....

```js
a("", "", ({}) => () => (a ? b : c));
```

Reply:
```console
# this one
>node bin/prettier test.js
a(
  "",
  "",
  ({}) =>
    () => (a ? b : c),
);

# stable
>node node_modules/prettier/bin-prettier.js test.js
a(
  "",
  "",
  ({}) =>
    () =>
      a ? b : c
);
```